_source/_posts/2021-08-02-fix-common-problems-cors.md

```diff
@@ -0,0 +1,530 @@
+
+From a developer's perspective, CORS is often a cause of much grief when it blocks network requests. CORS provides a number of different mechanisms for limiting JavaScript access to APIs. It is often not obvious which mechanism is blocking the request. We are going to build a simple web application that makes REST calls to a server in a different domain. We will deliberately make requests that the browser will block because of CORS policies and then show how to fix the issues. Let's get started!
+
+**NOTE**: The code for this project can be found on [GitHub](https://github.com/oktadev/okta-go-cors-example).
+
+## Prerequisites to Building a Go Application
+
+First things first, if you don't already have Go installed on your computer you will need to [download and install the Go Programming Language](https://golang.org/doc/install).
+
+Now, create a directory where all of our future code will live.
+
+```bash
+mkdir cors
+cd cors
+```
+
+Finally, we will make our directory a Go module and install the Gin package (a Go web framework) to implement a web server.
+
+```bash
+go mod init cors
+go get github.com/gin-gonic/gin
+go get github.com/gin-contrib/static
+```
+
+A file called `go.mod` will get created containing the dependencies.
+
+## How to Build a Simple Web Front End
+
+We are going to build a simple HTML and JavaScript front end and serve it from a web server written using Gin.
+
+First of all, create a directory called `frontend` and create a file called `frontend/index.html` with the following content:
+
+```html
+<html>
+    <head>
+        <meta charset="UTF-8" />
+        <title>Fixing Common Issues with CORS</title>
+        <script src="control.js" defer></script>
+    </head>
+    <body>
+        <h1>Fixing Common Issues with CORS</h1>
+        <div class="centred">
+            <textarea id="messages" name="messages" rows="10" cols="50">Messages</textarea><br/>
+            <form id="form1">
+                <input type="button" value="Get v1" onclick="onGet('v1')"/>
+                <input type="button" value="Get v2" onclick="onGet('v2')"/>
+            </form>
+        </div>
+    </body>
+</html>
+```
+
+The web page has a text area to display messages and a simple form with two buttons. When a button is clicked it calls the JavaScript function `onGet()` passing it a version number. The idea being that `v1` requests will always fail due to CORS issues, and `v2` will fix the issue.
+
+Next, create a file called `frontend/control.js` containing the following JavaScript:
+
+```javascript
+function onGet(version) {
+    const url = "http://localhost:8000/api/" + version + "/messages";
```

Comment:
JavaScript code is typically 2-space indented instead of 4. Probably doesn't matter if it's consistent throughout the post.

Reply:
I'll fix that in future posts, for now I did make sure all the js code was consistent